query/query.go

```diff
@@ -234,24 +238,31 @@ func NewResultBuilder(q Query) *ResultBuilder {
 	b := &ResultBuilder{
 		Query:  q,
 		Output: make(chan Result, bufSize),
+		closed: make(chan struct{}),
```

Comment:
Do we really need a `closed` channel, or would it essentially correspond to `ctx.Done()` ?

Reply:
This channel is needed in addition to the `ResultBuilder.ctx`. Cancelling the internal context (when Results is closed) tells the goroutine handling processing that it needs to exit and close the results (output) channel afterwards. -- That is the purpose of `ctx`.  The purpose of `closed` is to signal that the goroutine has exited and no more results are being read by the goroutine.

In other words...
- internal `ctx` signals "closing"
- closed, and therefore `Results.Done`, signals finished closing.

This is actual critical for some uses of query results when it is necessary to wait for the goroutine, started by `ResultsWithContext` to exit before doing something with items that were closed over by the function passes to `ResultsWithContext`.